File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T, U>(
  array: T[],
  predicate: (item: T) => boolean,
  fn: (item: T) => U
): U | undefined {
  for (const item of array) {
    if (predicate(item)) {
      return fn(item)
    }
  }
  return undefined
}

export function* powersGenerator(
  base: bigint
): Generator<bigint, void, unknown> {
  let power = 0n
  while (true) {
    yield base ** power
    power += 1n
  }
}

export async function meaningfulLineCount(filePath: string): Promise<number> {
  let validLineCount = 0

  const fileHandle = await open(filePath, "r")
  for await (const line of fileHandle.readLines()) {
    if (line.trim() && !line.trim().startsWith("#")) {
      validLineCount++
    }
  }
  return validLineCount
}

export type Shape = Box | Sphere

export interface Box {
  kind: "Box"
  width: number
  length: number
  depth: number
}

export interface Sphere {
  kind: "Sphere"
  radius: number
}

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return (
        2 *
        (shape.width * shape.length +
          shape.width * shape.depth +
          shape.length * shape.depth)
      )
    case "Sphere":
      return 4 * Math.PI * shape.radius ** 2
  }
}

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return shape.width * shape.length * shape.depth
    case "Sphere":
      return (4 / 3) * Math.PI * shape.radius ** 3
  }
}

export interface BinarySearchTree<T> {
  size(): number
  insert(value: T): BinarySearchTree<T>
  contains(value: T): boolean
  inorder(): Iterable<T>
  toString(): string
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0
  }

  insert(value: T): BinarySearchTree<T> {
    return new Node<T>(value, new Empty(), new Empty())
  }

  contains(value: T): boolean {
    return false
  }

  *inorder(): Iterable<T> {}

  toString(): string {
    return "()"
  }
}

class Node<T> implements BinarySearchTree<T> {
  private value: T
  private left: BinarySearchTree<T>
  private right: BinarySearchTree<T>
  private treeSize: number

  constructor(value: T, left: BinarySearchTree<T>, right: BinarySearchTree<T>) {
    this.left = left
    this.right = right
    this.value = value
    this.treeSize = 1 + left.size() + right.size()
  }

  size(): number {
    return this.treeSize
  }

  insert(value: T): BinarySearchTree<T> {
    if (value < this.value) {
      return new Node(this.value, this.left.insert(value), this.right)
    } else if (value > this.value) {
      return new Node(this.value, this.left, this.right.insert(value))
    } else {
      return this
    }
  }

  contains(value: T): boolean {
    if (value < this.value) {
      return this.left.contains(value)
    } else if (value > this.value) {
      return this.right.contains(value)
    } else {
      return true
    }
  }

  *inorder(): Iterable<T> {
    yield* this.left.inorder()
    yield this.value
    yield* this.right.inorder()
  }

  toString(): string {
    return `(${this.left.toString().replace("()", "")}${this.value}${this.right
      .toString()
      .replace("()", "")})`
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

let first_then_apply array predicate consumer =
  match List.find_opt predicate array with
  | Some x -> consumer x
  | None -> None

let powers_generator base = 
  let rec generate_power power () = 
    Seq.Cons (power, generate_power (power * base))
  in
  generate_power 1
  
let meaningful_line_count filename =
  let meaningful_line line = 
    let trimmed = String.trim line in
    String.length trimmed > 0 && not (String.starts_with ~prefix:"#" trimmed)
  in

  let the_file = open_in filename in
  let finally () = close_in the_file in

  let rec count_lines count =
    try
      let line = input_line the_file in 
      if meaningful_line line then
        count_lines (count + 1)
      else
        count_lines count
    with
    | End_of_file -> count
  in
  Fun.protect ~finally (fun () -> count_lines 0)

type shape = 
  | Sphere of float
  | Box of float * float * float

let surface_area s =
  match s with 
  | Sphere radius -> 4.0 *. Float.pi *. (radius**2.0)
  | Box (length, width, height) -> 2.0 *. length *. width +. 2.0 *. length *. height +. 2.0 *. height *. width

let volume s = 
  match s with 
  | Sphere radius -> (4.0 /. 3.0) *. Float.pi *. (radius**3.0)
  | Box (length, width, height) -> length *. width *. height

(* Write your binary search tree implementation here *)
type 'a binary_search_tree = 
	| Empty
	| Node of 'a binary_search_tree * 'a * 'a binary_search_tree

let rec size tree =
  match tree with
  | Empty -> 0
  | Node (left, _, right) -> 1 + size left + size right

let rec contains data tree =
  match tree with
  | Empty -> false
  | Node (left, value, right) -> 
    if data = value then
      true
    else if data < value then
      contains data left 
    else 
      contains data right

let rec insert data tree =
  match tree with 
  | Empty -> Node(Empty, data, Empty)
  | Node (left, value, right) -> 
    if data < value then 
      Node (insert data left, value, right)
    else if data > value then 
      Node (left, value, insert data right)
    else
      Node (left, value, right)

let rec inorder tree =
  match tree with
  | Empty -> []
  | Node (left, value, right) -> inorder left @ [value] @ inorder right